### PR TITLE
Fix flow calculation

### PR DESCRIPTION
**Pull request by jlebar**
<git-pr-chain>

#### Commits in this PR
1. Fix flow calculation.
    
    Instead of
    
      f(pressure_inhalation) - f(pressure_exhalation)
    
    it should be
    
      f(pressure_inhalation - pressure_exhalation),
    
    where f is pressure_delta_to_volumetric_flow.
    
    Sadly no unit tests yet, because pid.cpp can't be tested until the PID
    library it depends on is HAL-ified.
1. Rename some identifiers in sensors module.
    
    In the previous commit we identified a bug; we did f(x) - f(y) instead
    of f(x-y).  My guess is that maybe this was caused/exacerbated by a
    function name (that I suggested) being too long?  Try to shorten the
    name.

#### [PR chain](https://github.com/jlebar/git-pr-chain)
1. #172 Cosmetic and functional improvements to pid.cpp
1. 👉 #173 Fix flow calculation 👈 **YOU ARE HERE**

</git-pr-chain>

### CODE REVIEW

File: controller/src/pid.cpp [hunk 1]
```diff
@@ -139,18 +139,18 @@ void pid_execute(const VentParams &params, SensorReadings *readings) {
...
-      (pressure_delta_to_volumetric_flow(
-           get_pressure_reading(PressureSensors::INHALATION_PIN)) -
-       pressure_delta_to_volumetric_flow(
-           get_pressure_reading(PressureSensors::EXHALATION_PIN))) *
-      1000.0f * 1000.0f * 60.0f;
+  float flow = // [m^3/s]
+      pressure_delta_to_flow(
+          get_pressure_reading(PressureSensors::INHALATION_PIN) -
+          get_pressure_reading(PressureSensors::EXHALATION_PIN));
+  readings->flow_ml_per_min = flow * 1000.0f * 1000.0f * 60.0f;
```
Comment (verityRF) [inline controller/src/pid.cpp:155]: This is cleaner, looks good.

File: controller/test/sensor_tests/SensorTests.cpp [hunk 2]
```diff
@@ -164,42 +164,42 @@ TEST(SensorTests, FullScaleReading) {
...
-  float volumFlow = pressure_delta_to_volumetric_flow(1.0e-7f);
+  float volumFlow = pressure_delta_to_flow(1.0e-7f);
   // 1e-7 kPa differential pressure should result in 3.07e-7 [m^3/s] of Q
   EXPECT_NEAR(volumFlow, 3.07e-7f, COMPARISON_TOLERANCE_FLOW);
 }
 
 TEST(SensorTests, TestLargeVolumetricFlowCalculation) {
   sensors_init();
-  float volumFlow = pressure_delta_to_volumetric_flow(100.0f);
+  float volumFlow = pressure_delta_to_flow(100.0f);
```
Comment (verityRF) [inline controller/test/sensor_tests/SensorTests.cpp:195]: Nit: I put volumetric in implicitly so as to be abundantly clear that this is volumetric flow, not flow velocity, which is different.

File: controller/src/pid.cpp (hunk 1 above)
Comment (jlebar) [reply controller/src/pid.cpp:155]: Just to check -- this *is* correct, right?  Like, I think I'm fixing a bug, just want to be sure I'm not introducing one.

File: controller/test/sensor_tests/SensorTests.cpp (hunk 2 above)
Comment (jlebar) [reply controller/test/sensor_tests/SensorTests.cpp:195]: Oh, you wouldn't call the other one "velocity"?

Comment (verityRF) [reply controller/test/sensor_tests/SensorTests.cpp:195]: They both are a velocity, but there are subtle differences, they are not equivalent. Flow velocity is used to calculate the volumetric flow.

File: controller/src/pid.cpp (hunk 1 above)
Comment (verityRF) [reply controller/src/pid.cpp:155]: Oh dear, this is actually incorrect. Good call. f(x-y) =/= f(x) - f(y) in this case. Intuitively it *makes* sense for some reason for this setup, but mathematically it's wrong. 

If you look at pressure_delta_to_volumetric, the equation that translates delta pressure to Q is essentially f(deltaP) = constant * sqrt(deltaP). The square root makes f(x-y) =/= f(x) - f(y).

Comment (jlebar) [reply controller/src/pid.cpp:155]: > f(x-y) =/= f(x) - f(y) in this case.

Agreed.

> The original is correct, this is not.

The function is called "pressure_delta_to_volumetric_flow" but when we do `f(x) - f(y)` aren't we passing in absolute pressures `x` and `y`?  Whereas `f(x-y)` is a pressure delta?

File: controller/test/sensor_tests/SensorTests.cpp (hunk 2 above)
Comment (jlebar) [reply controller/test/sensor_tests/SensorTests.cpp:195]: Are you telling me my physics teacher LIED to me when she said that velocity is defined as dPosition/dT (and therefore has units m/s)?  Like, volumetric flow has units m^3/s, which is different!  (Whereas flow velocity is indeed a vector field with units m/s.)

Destroying my childhood.  :(

**Comment by jlebar** [comment]
> f(delta_pressure_inhalation) - f(delta_pressure_exhalation) is correct.

Finally got it, thanks for being patient with me.

I will send a PR to clarify that the patient pressure sensor reads an absolute pressure (correct?), whereas these return a pressure difference over the venturi.

**Comment by jlebar** [comment]
I'll close this and try again.  Thanks, Kyle.

**Comment by verityRF** [comment]
Yeah I believe your confusion is with what each pressure sensor does. The patient sensor is relative to atmosphere, the inhalation/exhalation are different part numbers and measure across the venturi, so they individually are already a differential pressure - don't need to subtract.

**Pull request closed without merging**